### PR TITLE
Incremental QOS config updater test update

### PR DESCRIPTION
**Pull request by isabelmsft**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911

#### What is the motivation for this PR?

[End to End test support for Generic Updater apply-patch](https://github.com/Azure/SONiC/blob/master/doc/config-generic-update-rollback/SONiC_Generic_Config_Update_and_Rollback_Design.md)
Update qos config values to update based on Mellanox SPC1 buffer calculations 

#### How did you do it?
Calculate proper buffer field values based, integrate into config updater JSON patch

#### How did you verify/test it?
Tested on msn2700 devices, all tests pass

#### Any platform specific information?
For Mellanox msn2700 

#### Supported testbed topology if it's a new test case?
t0, t1

### CODE REVIEW

File: tests/generic_config_updater/test_incremental_qos.py [hunk 1]
```diff
@@ -25,69 +38,185 @@ def ensure_dut_readiness(duthost):
...
+        duthost: DUT host object
+        tbinfo: information about the running testbed
 
-def prepare_configdb_field(duthost, configdb_field, value):
+    Returns:
+        uplink count, downlink count
+
+    """
+    config_db_json = json.loads(duthost.shell("show runningconfig all")["stdout"])
+    device_neighbor_metadata = config_db_json["DEVICE_NEIGHBOR_METADATA"]
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_incremental_qos.py:67]: >device_neighbor_metadata [](http://example.com/codeflow?start=4&length=24)

You may reuse `duthost.config_facts`
ref: [https://github.com/Azure/sonic-mgmt/blob/3a9720789b4cd4527865b0053795e98c14ccfd2c/tests/test_nbr_health.py#L81](https://github.com/Azure/sonic-mgmt/blob/3a9720789b4cd4527865b0053795e98c14ccfd2c/tests/test_nbr_health.py#L81)

File: tests/generic_config_updater/test_incremental_qos.py [hunk 2]
```diff
@@ -25,69 +38,185 @@ def ensure_dut_readiness(duthost):
...
+    topo = tbinfo['topo']['name']
+
+    if "t1" in topo:
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_incremental_qos.py:70]: >topo [](http://example.com/codeflow?start=15&length=4)

To be more specific, is it better to check `topo.startswitch('topo_t1')`?

File: tests/generic_config_updater/test_incremental_qos.py [hunk 3]
```diff
@@ -25,69 +38,185 @@ def ensure_dut_readiness(duthost):
...
+
+    for neighbor in neighbor_set:
+        interface = neighbor_to_interface_map[neighbor]
+
+        cable_length = duthost.shell('sonic-db-cli CONFIG_DB hget "CABLE_LENGTH|AZURE" {}'.format(interface))['stdout']
+        port_speed = duthost.shell('sonic-db-cli CONFIG_DB hget "PORT|{}" speed'.format(interface))['stdout']
+
+        expected_profile = 'pg_lossless_{}_{}_profile'.format(port_speed, cable_length)
+
+        xoff = int(duthost.shell('redis-cli hget "BUFFER_PROFILE_TABLE:{}" xoff'.format(expected_profile))['stdout'])
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_incremental_qos.py:130]: >redis-cli [](http://example.com/codeflow?start=34&length=9)

`redis-cli` will read AppDB (database 0) by default. I think you need `sonic-db-cli CONFIG_DB`.

File: tests/generic_config_updater/test_incremental_qos.py [hunk 4]
```diff
@@ -25,69 +38,185 @@ def ensure_dut_readiness(duthost):
...
+        duthost: DUT host object
+        configdb_field: config db field under test
+        value: expected value of configdb_field
+    """
+    def _confirm_value_in_asic_db():
+        if "ingress_lossless_pool" in configdb_field:
+            buffer_pool = "ingress_lossless_pool"
+        elif "egress_lossy_pool" in configdb_field:
+            buffer_pool = "egress_lossy_pool"
+        oid = duthost.shell('redis-cli -n 2 HGET COUNTERS_BUFFER_POOL_NAME_MAP {}'.format(buffer_pool))["stdout"]
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_incremental_qos.py:191]: >redis-cli -n 2 [](http://example.com/codeflow?start=29&length=14)

`redis-cli` is deprecated. Suggest use `sonic-db-cli ASIC_DB`

File: tests/generic_config_updater/test_incremental_qos.py [hunk 5]
```diff
@@ -8,14 +8,27 @@
 from tests.common.helpers.dut_utils import verify_orchagent_running_or_assert
 from tests.generic_config_updater.gu_utils import apply_patch, expect_op_success, expect_res_success, expect_op_failure
 from tests.generic_config_updater.gu_utils import generate_tmpfile, delete_tmpfile
+from tests.generic_config_updater.gu_utils import create_checkpoint, delete_checkpoint, rollback_or_reload
 
 pytestmark = [
-    pytest.mark.topology('t0'),
+    pytest.mark.topology('t0', 't1'),
     pytest.mark.asic('mellanox')
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_incremental_qos.py:15]: >mellanox [](http://example.com/codeflow?start=22&length=8)

Is it safe to run this test on mellanox switch but not 2700?

**Comment by qiluo-msft** [comment]
Could you fix valid LGTM alerts in the build checkers?

File: tests/generic_config_updater/test_incremental_qos.py (hunk 5 above)
Comment (isabelmsft) [reply tests/generic_config_updater/test_incremental_qos.py:14]: These formula are only applicable to 2700

File: tests/generic_config_updater/test_incremental_qos.py (hunk 2 above)
Comment (isabelmsft) [reply tests/generic_config_updater/test_incremental_qos.py]: Changed to `topo.startswitch('topo_t1-')`

Comment (qiluo-msft) [reply tests/generic_config_updater/test_incremental_qos.py]: i think there is a topo called `topo_t1`, and we should test it.

File: tests/generic_config_updater/test_incremental_qos.py [hunk 6]
```diff
@@ -25,69 +37,188 @@ def ensure_dut_readiness(duthost):
...
+        interface = neighbor_to_interface_map[neighbor]
+
+        cable_length = duthost.shell('sonic-db-cli CONFIG_DB hget "CABLE_LENGTH|AZURE" {}'.format(interface))['stdout']
+        port_speed = duthost.shell('sonic-db-cli CONFIG_DB hget "PORT|{}" speed'.format(interface))['stdout']
+
+        expected_profile = 'pg_lossless_{}_{}_profile'.format(port_speed, cable_length)
+
+        xoff = int(duthost.shell('redis-cli hget "BUFFER_PROFILE_TABLE:{}" xoff'.format(expected_profile))['stdout'])
+        xon = int(duthost.shell('redis-cli hget "BUFFER_PROFILE_TABLE:{}" xon'.format(expected_profile))['stdout'])
+        pg_headroom = (xoff + xon) / 1024
```
Comment (qiluo-msft) [inline tests/generic_config_updater/test_incremental_qos.py:131]: >/ [](http://example.com/codeflow?start=35&length=1)

LGTM reports a warning here.
The main concern is that the behavior of `/` is different in python3 vs python2. And we can improve the code to make it behavior the same way.

You may find some solutions [https://stackoverflow.com/a/15633893/2514803](https://stackoverflow.com/a/15633893/2514803). I prefer `from __future__ import division` with a `int()` conversion.

File: tests/generic_config_updater/test_incremental_qos.py (hunk 5 above)
Comment (isabelmsft) [reply tests/generic_config_updater/test_incremental_qos.py:14]: @wangxin Is there a marker we can add for 2700?

Comment (yxieca) [reply tests/generic_config_updater/test_incremental_qos.py:14]: Do you only want this test to run on 2700? If so, you can add a pytest_requires condition in your test code entrance to only allow 2700. But why?

Comment (isabelmsft) [reply tests/generic_config_updater/test_incremental_qos.py:15]: Yes, the Mellanox SPC1 buffer calculations only apply to 2700 devices. @neethajohn may confirm/correct me

And thanks, I added `pytest_require` to check for 2700 in hwsku

File: tests/generic_config_updater/test_incremental_qos.py (hunk 2 above)
Comment (isabelmsft) [reply tests/generic_config_updater/test_incremental_qos.py:70]: Got access to t1-lag topology, and `startswith` also incorrectly skipped this testbed. I changed back to check for `"t1" in topo`

**Review by qiluo-msft — DISMISSED** [review]

**Review by qiluo-msft — DISMISSED** [review]

**Comment by qiluo-msft** [comment]
The PR includes so many files due to recently merge. Is it expected?

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**